### PR TITLE
S3 Object Lock Demo

### PR DESCRIPTION
**Pull request by ravinitp**
This Go project is a demonstration of a simple command-line tool to manage object locking in an AWS S3 bucket. It allows users to acquire and release locks on a specified file in an S3 bucket. This project is designed to showcase the concept of object locking in AWS S3 and is a minimalistic implementation for educational purposes.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		&s3.DeleteObjectInput{
+			Key:    aws.String(path),
+			Bucket: aws.String(bucket),
+		})
+	if err != nil {
+		fmt.Printf("[WARNING] It seems that some one is trying to aquire the lock simultaneously.")
+	}
+	fmt.Printf("delete Response   %v\n", deleteResponse)
+	// Read version of state file
+	getResponse, err := sdk.GetObject(&s3.GetObjectInput{
```
Comment (sansmoraxz) [inline main.go:83]: Should this err not be checked?

**Comment by sansmoraxz** [comment]
Seems promising

File: main.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		Key:    aws.String(lockFileName(path)),
+		Bucket: aws.String(bucket),
+	})
+	fmt.Printf("read Response   %v\n", getResponse)
+
+	if *copyResponse.VersionId != *getResponse.VersionId {
```
Comment (sansmoraxz) [inline main.go:89]: Be aware that this will segfault if bucket versioning is not enabled. Probably should be checked earlier somewhere.

**Comment by sansmoraxz** [comment]
Race still occurs while testing.
Used a bit larger file (the AWS CLI), with random delays.

```go
time.Sleep(time.Duration(rand.Int31n(5)) * time.Second)
```


Occasionally only the lock file remains.

File: main.go (hunk 2 above)
Comment (ravinitp) [reply main.go:89]: > Be aware that this has a change to segfault if bucket versioning is not enabled. Probably should be checked earlier somewhere.

thanks for ur feedback. I will take care while implementing it.

File: main.go (hunk 1 above)
Comment (ravinitp) [reply main.go:83]: thanks for ur feedback. I will take care while implementing it.

**Comment by ravinitp** [comment]
> Race still occurs while testing. Used a bit larger file (the AWS CLI), with random delays.
> 
> ```go
> time.Sleep(time.Duration(rand.Int31n(5)) * time.Second)
> ```
> 
> Occasionally only the lock file remains.
@sansmoraxz 
I would like to hear more about this issue. Please add steps to reproduce.
specially where did you add sleep statement.

Although me and Marcin had discussion and decided to use a small dummy object to lock state file. We are waiting for AWS team to revert on query.

**Comment by sansmoraxz** [comment]
I was using your test case. Still investigating, I might be wrong but your coroutines may not be synced properly.

This brings into another concern what happens if the terraform binary crashes. The lock will be still attached.

**Comment by ravinitp** [comment]
> sansmoraxz

> This brings into another concern what happens if the terraform binary crashes. The lock will be still attached.

This is possible with other locks as well. But this can be resolved with -forcelock.

**Pull request merged**